# Mistral Training Samples - Full Prompts and Responses

This notebook demonstrates the training pipeline by generating climate data samples and showing the complete prompts (including tabulated statistics) and Mistral's responses.

The pipeline:
1. Loads climate data from Zarr files
2. Selects random geographic locations
3. Computes statistics (min/max/mean/gradient/rotation/divergence)
4. Formats statistics as tables
5. Generates prompts for Mistral
6. Gets real Mistral-7B-Instruct-v0.3 responses

In [15]:
# Import required libraries
import sys
from pathlib import Path

# Add parent directory to path
sys.path.insert(0, str(Path.cwd().parent.parent))

from multimodal_aifs.training import ClimateTextDataLoader


## Initialize the DataLoader

This will load the Ministral-8B-Instruct-2410 model and prepare the training pipeline.

In [16]:
# Initialize the dataloader
zarr_path = Path.cwd().parent.parent / "test_aifs_large.zarr"

print(f"Loading from: {zarr_path}")
print("\nNote: This will load the Ministral-8B-Instruct-2410 model (~16GB)")
print("Model will be loaded in float16 on MPS (Apple Silicon GPU) for efficiency\n")

dataloader = ClimateTextDataLoader(
    zarr_paths=str(zarr_path),
    samples_per_epoch=3,
    cache_mistral_model=True,
    device="cpu",  # Will auto-detect and use MPS if available
    seed=42
)

print(f"\n✓ DataLoader initialized successfully")


Loading from: /Users/arigazzi/Documents/DeepLearning/LLM for climate/HPE-LLM4Climate/test_aifs_large.zarr

Note: This will load the Ministral-8B-Instruct-2410 model (~16GB)
Model will be loaded in float16 on MPS (Apple Silicon GPU) for efficiency

Initializing climate-text dataloader components...
Loading tokenizer: mistralai/Ministral-8B-Instruct-2410
Loading Mistral model: mistralai/Ministral-8B-Instruct-2410
  Loading model in float16 for memory efficiency...


Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

  ✓ Mistral model loaded in float16 mode on mps:0
  ✓ Loaded test_aifs_large.zarr: 2 timesteps
DataLoader initialized:
  - Zarr files: 1
  - Samples per epoch: 3
  - Device: cpu

✓ DataLoader initialized successfully

✓ DataLoader initialized successfully


## Generate Training Samples

Generate 3 samples with real climate data and Ministral responses.

**Note:** Using `max_response_length=2048` to get full-length responses instead of the default 512 tokens used during training.

In [17]:
# Generate samples with full-length responses (2048 tokens instead of default 512)
print("Generating 3 training samples...\n")
samples = dataloader.get_sample_preview(num_samples=3, max_response_length=2048)
print(f"\n✓ Generated {len(samples)} samples")


Generating 3 training samples...



TypeError: ClimateTextDataLoader.get_sample_preview() got an unexpected keyword argument 'max_response_length'

---
# Sample 1: Pacific Ocean

## Location Information

In [ ]:
sample1 = samples[0]

print(f"Location: {sample1.location.name}")
print(f"Type: {sample1.location.location_type}")
print(f"Coordinates: {sample1.location.center_lat:.2f}°N, {sample1.location.center_lon:.2f}°E")
print(f"Masked grid points: {sample1.mask.sum()} / {sample1.mask.shape[0]}")
print(f"Timesteps: {sample1.timestep_1_index} → {sample1.timestep_2_index}")


Location: Pacific Ocean
Type: body_of_water
Coordinates: 0.00°N, -160.00°E
Masked grid points: 27 / 542080
Timesteps: 0 → 1


## Full Prompt (with Statistics Table)

In [ ]:
print(sample1.prompt)


Analyze the weather conditions in the Pacific Ocean based on the following climate data:

Variable             Min          Max          Mean         Gradient     Rotation     Divergence  
z_50                     -1.3397      2.2015      0.0135  7.0153e-01 N/A          N/A         
z_100                    -1.5327      1.3373     -0.0709  8.5292e-01 N/A          N/A         
z_150                    -1.1166      1.5978      0.3164  8.5503e-01 N/A          N/A         
z_200                    -1.9587      1.4104     -0.0133  9.0366e-01 N/A          N/A         
z_250                    -1.8337      1.2816     -0.2400  1.0177e+00 N/A          N/A         
z_300                    -0.7864      1.6797      0.1804  7.6638e-01 N/A          N/A         
z_400                    -1.7335      1.5343     -0.0795  9.6590e-01 N/A          N/A         
z_500                    -1.2654      2.1270      0.1025  8.0913e-01 N/A          N/A         
z_600                    -1.7550      1.4975      0

## Mistral Response

In [ ]:
print(sample1.llm_response)


vided climate data, here's an analysis of the weather conditions in the Pacific Ocean:

### Geopotential Heights (z)
- **z_50 to z_850**: The geopotential heights at these levels show a general decrease from 50 hPa to 850 hPa, indicating a slight increase in altitude with increasing pressure. This suggests a stable atmospheric profile, with a potential for a weak or no vertical wind shear.

- **z_100 to z_250**: The geopotential heights at these levels show a slight increase in altitude with increasing pressure, indicating a stable atmospheric profile. This is consistent with the lower levels.

- **z_300 to z_700**: The geopotential heights at these levels show a decrease in altitude with increasing pressure, suggesting a stable atmosphere. However, the change in geopotential height is relatively small, indicating minimal vertical wind shear.

### Temperature (2t)
- The temperature (2t) data shows a mean temperature of 0.0812 K. This suggests that the atmosphere is slightly warmer at t

---
# Sample 2: Australia

## Location Information

In [ ]:
sample2 = samples[1]

print(f"Location: {sample2.location.name}")
print(f"Type: {sample2.location.location_type}")
print(f"Coordinates: {sample2.location.center_lat:.2f}°N, {sample2.location.center_lon:.2f}°E")
print(f"Masked grid points: {sample2.mask.sum()} / {sample2.mask.shape[0]}")
print(f"Timesteps: {sample2.timestep_1_index} → {sample2.timestep_2_index}")


Location: Australia
Type: continent
Coordinates: -25.00°N, 135.00°E
Masked grid points: 18204 / 542080
Timesteps: 0 → 1


## Full Prompt (with Statistics Table)

In [ ]:
print(sample2.prompt)


Analyze the weather conditions in Australia based on the following climate data:

Variable             Min          Max          Mean         Gradient     Rotation     Divergence  
z_50                     -2.7681      2.8975      0.0057  7.9550e-01 N/A          N/A         
z_100                    -2.8709      2.7760      0.0021  7.9291e-01 N/A          N/A         
z_150                    -2.7235      3.2172     -0.0079  8.0058e-01 N/A          N/A         
z_200                    -2.7311      2.4942     -0.0028  8.0278e-01 N/A          N/A         
z_250                    -2.8652      2.9703      0.0102  8.0196e-01 N/A          N/A         
z_300                    -2.8841      2.6852     -0.0029  8.0091e-01 N/A          N/A         
z_400                    -2.7695      2.7516      0.0049  7.9240e-01 N/A          N/A         
z_500                    -2.9035      2.7941     -0.0017  7.9658e-01 N/A          N/A         
z_600                    -2.5274      2.7312     -0.0033  8

## Mistral Response

In [ ]:
print(sample2.llm_response)


vided climate data, we can infer several aspects of the weather conditions in Australia. However, note that the data provided includes geopotential heights, not direct temperature and wind speeds. Let's break down the key weather elements:

### Temperature
- **Geopotential Height (z):** Geopotential height is a measure of the height above a reference level (typically sea level) where the air pressure is equal to the standard atmosphere pressure. The values range from negative to positive, where negative values indicate heights below the reference level, and positive values indicate heights above the reference level.

  - **Mean and Gradient:** The mean geopotential heights vary from -2.9664 (at 1000 hPa) to 3.1106 (at z). This suggests a general range of heights where the air is either above or below the reference level. The gradients (differences between min and max values) are mostly small, indicating slight changes in height.

  - **Notable Heights:**
    - **z_50 (50 hPa):** -2.768

---
# Sample 3: Spain

## Location Information

In [ ]:
sample3 = samples[2]

print(f"Location: {sample3.location.name}")
print(f"Type: {sample3.location.location_type}")
print(f"Coordinates: {sample3.location.center_lat:.2f}°N, {sample3.location.center_lon:.2f}°E")
print(f"Masked grid points: {sample3.mask.sum()} / {sample3.mask.shape[0]}")
print(f"Timesteps: {sample3.timestep_1_index} → {sample3.timestep_2_index}")


Location: Spain
Type: country
Coordinates: 40.50°N, -3.70°E
Masked grid points: 24 / 542080
Timesteps: 0 → 1


## Full Prompt (with Statistics Table)

In [ ]:
print(sample3.prompt)


Analyze the weather conditions in Spain based on the following climate data:

Variable             Min          Max          Mean         Gradient     Rotation     Divergence  
z_50                     -2.6548      1.6031      0.0747  9.2899e-01 N/A          N/A         
z_100                    -1.1166      1.1380      0.0438  6.8228e-01 N/A          N/A         
z_150                    -1.0429      1.1632     -0.1701  5.7841e-01 N/A          N/A         
z_200                    -1.9419      1.8437     -0.0777  9.7737e-01 N/A          N/A         
z_250                    -1.1502      1.0752      0.0324  7.9855e-01 N/A          N/A         
z_300                    -1.7340      1.1946      0.0193  8.8495e-01 N/A          N/A         
z_400                    -1.4171      0.7736     -0.2255  7.5331e-01 N/A          N/A         
z_500                    -1.9384      1.3010     -0.0845  1.0566e+00 N/A          N/A         
z_600                    -1.3503      0.8762      0.1179  7.072

## Mistral Response

In [ ]:
print(sample3.llm_response)


a appears to be geopotential height and related atmospheric variables at different pressure levels. To describe the weather conditions based on this data, let's break down the information:

### Temperature
- **Geopotential Height (z)**: This is a measure of the height of a given pressure level above the standard sea level. The geopotential height data (z_50 to z_1000) indicates the altitude of various pressure levels.
- **Temperature (2t)**: This variable represents the temperature at 2 meters above the ground. The mean value is -0.0151°C, and the gradient indicates a slight decrease in temperature with height, suggesting a stable atmospheric condition.

### Wind
- **Wind Components (10u, 10v)**: These represent the zonal (east-west) and meridional (north-south) wind components at 10 meters above the ground. The mean wind speed is -0.2518 m/s for 10u and 0.0044 m/s for 10v, indicating a slight wind from the west-southwest.
- **Wind Rotation**: The gradient indicates a slight clockwise 

---
# Summary

## Sample Statistics

In [ ]:
import pandas as pd

summary_data = {
    'Sample': ['1', '2', '3'],
    'Location': [s.location.name for s in samples],
    'Type': [s.location.location_type for s in samples],
    'Masked Points': [s.mask.sum().item() for s in samples],
    'Prompt Length': [len(s.prompt) for s in samples],
    'Response Length': [len(s.llm_response) for s in samples],
}

summary_df = pd.DataFrame(summary_data)
print(summary_df.to_string(index=False))


Sample      Location          Type  Masked Points  Prompt Length  Response Length
     1 Pacific Ocean body_of_water             27           3027             1980
     2     Australia     continent          18204           3019             1941
     3         Spain       country             24           3015             2331


## Tensor Shapes Verification

In [ ]:
print("Verifying tensor shapes for AIFS model compatibility:\n")

for i, sample in enumerate(samples, 1):
    print(f"Sample {i}:")
    print(f"  climate_tensor_t1: {sample.climate_tensor_t1.shape}")
    print(f"  climate_tensor_t2: {sample.climate_tensor_t2.shape}")
    print(f"  aifs_input_tensor: {sample.aifs_input_tensor.shape}")
    print(f"  mask: {sample.mask.shape}")
    print()


Verifying tensor shapes for AIFS model compatibility:

Sample 1:
  climate_tensor_t1: torch.Size([542080, 103])
  climate_tensor_t2: torch.Size([542080, 103])
  aifs_input_tensor: torch.Size([1, 2, 1, 542080, 103])
  mask: torch.Size([542080])

Sample 2:
  climate_tensor_t1: torch.Size([542080, 103])
  climate_tensor_t2: torch.Size([542080, 103])
  aifs_input_tensor: torch.Size([1, 2, 1, 542080, 103])
  mask: torch.Size([542080])

Sample 3:
  climate_tensor_t1: torch.Size([542080, 103])
  climate_tensor_t2: torch.Size([542080, 103])
  aifs_input_tensor: torch.Size([1, 2, 1, 542080, 103])
  mask: torch.Size([542080])



---
# Notes

- All samples use **real climate data** from Zarr files
- **Real Ministral-8B-Instruct-2410** responses (no synthetic data)
- Model runs in **float16** on **MPS** (Apple Silicon GPU)
- Statistics include: min, max, mean, gradient, rotation, divergence
- Each sample covers different geographic scales:
  - **Body of water** (Pacific Ocean): 27 grid points
  - **Continent** (Australia): 18,204 grid points
  - **Country** (Spain): 24 grid points